DOCSUP-8324: Document the support of replicas and shards in MySQL/PostgreSQL

sevirov opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):

- Documentation for #22217

sevirov committed:
Create ExternalDistributed.md

Задокументировал новый табличный движок ExternalDistributed.

sevirov committed:
Update ExternalDistributed.md

Внес небольшие поправки.

sevirov committed:
Fix See Also

Поправил ссылки.

kssenii commented on docs/en/engines/table-engines/integrations/ExternalDistributed.md:
```diff
@@ -0,0 +1,70 @@
+-   `table` — Remote table name.
+-   `user` — User name.
+-   `password` — User password.
+-   `schema` — Non-default table schema. Optional.
```
`schema` parameter was not added, but should have been though.

kssenii commented on docs/en/engines/table-engines/integrations/ExternalDistributed.md:
```diff
@@ -0,0 +1,70 @@
+
+## Implementation Details {#implementation-details}
+	
+Supports multiple replicas that must be listed by a character `|`. For example:
```
And shards are listed with `,`.

kssenii commented on docs/en/engines/table-engines/integrations/ExternalDistributed.md:
````diff
@@ -0,0 +1,70 @@
+CREATE TABLE test_shards (id UInt32, name String, age UInt32, money UInt32) ENGINE = ExternalDistributed('MySQL', `mysql{1|2}:3306,mysql{3|4}:3306`, 'clickhouse', 'test_replicas', 'root', 'clickhouse');
+```
+
+When specifying replicas, one of the available replicas will be selected for each of the shards when reading. You can configure the algorithm for load balancing (the preference for which replica to access) – see the [load_balancing](../../../operations/settings/settings.md#settings-load_balancing) setting. If the connection with the server is not established, there will be an attempt to connect w ...
````
> You can configure the algorithm for load balancing (the preference for which replica to access) – see the [load_balancing]

No, it is not configurable via `load_balancing` setting.

> there will be an attempt to connect with a short timeout.

There is not timeout.

>  a remote server might accept the connection, but might not work

Questionable - if connection is accepted, then it is actually connected, but might disconnect any moment after.

kssenii commented on docs/en/engines/table-engines/integrations/ExternalDistributed.md:
```diff
@@ -0,0 +1,70 @@
+
+When specifying replicas, one of the available replicas will be selected for each of the shards when reading. You can configure the algorithm for load balancing (the preference for which replica to access) – see the [load_balancing](../../../operations/settings/settings.md#settings-load_balancing) setting. If the connection with the server is not established, there will be an attempt to connect w ...
+
+You can specify just one of the shards (in this case, query processing should be called remote, rather than distributed) or up to any number of shards. In each shard, you can specify from one to any number of replicas. You can specify a different number of replicas for each shard.
```
> You can specify just one of the shards 

In this case there is no point to use this table engine as it will be the same as using just MySQL/PostgreSQL storages directly.

kssenii commented on docs/en/engines/table-engines/integrations/ExternalDistributed.md:
```diff
@@ -0,0 +1,70 @@
+
+You can specify just one of the shards (in this case, query processing should be called remote, rather than distributed) or up to any number of shards. In each shard, you can specify from one to any number of replicas. You can specify a different number of replicas for each shard.
+
+Each shard can have a weight defined in the config file. By default, the weight is equal to one. Data is distributed across shards in the amount proportional to the shard weight. For example, if there are two shards and the first has a weight of 9 while the second has a weight of 10, the first will be sent 9 / 19 parts of the rows, and the second will be sent 10 / 19.
```
> Each shard can have a weight defined in the config file.

Not true - there are no shard weights.

kssenii commented on docs/en/engines/table-engines/integrations/ExternalDistributed.md:
```diff
@@ -0,0 +1,70 @@
+
+Each shard can have a weight defined in the config file. By default, the weight is equal to one. Data is distributed across shards in the amount proportional to the shard weight. For example, if there are two shards and the first has a weight of 9 while the second has a weight of 10, the first will be sent 9 / 19 parts of the rows, and the second will be sent 10 / 19.
+
+To select the shard that a row of data is sent to, the sharding expression is analyzed, and its remainder is taken from dividing it by the total weight of the shards. The row is sent to the shard that corresponds to the half-interval of the remainders from `prev_weight` to `prev_weights + weight`, where `prev_weights` is the total weight of the shards with the smallest number, and `weight` is the ...
```
This line and below about shard weights cannot be applied as there are no shard weights.

sevirov committed:
Translate to Russian

Выполнил перевод на русский язык.

olgarev commented on docs/en/engines/table-engines/integrations/ExternalDistributed.md:
````diff
@@ -0,0 +1,54 @@
+CREATE TABLE test_shards (id UInt32, name String, age UInt32, money UInt32) ENGINE = ExternalDistributed('MySQL', `mysql{1|2}:3306,mysql{3|4}:3306`, 'clickhouse', 'test_replicas', 'root', 'clickhouse');
+```
+
+When specifying replicas, one of the available replicas will be selected for each of the shards when reading. If the connection failed, the next replica will be selected, and so on for all the replicas. If the connection attempt failed for all the replicas, the attempt will be repeated the same way several times.
````
```suggestion
When specifying replicas, one of the available replicas is selected for each of the shards when reading. If the connection fails, the next replica is selected, and so on for all the replicas. If the connection attempt fails for all the replicas, the attempt is repeated the same way several times.
```

olgarev commented on docs/en/engines/table-engines/integrations/ExternalDistributed.md:
```diff
@@ -0,0 +1,54 @@
+
+When specifying replicas, one of the available replicas will be selected for each of the shards when reading. If the connection failed, the next replica will be selected, and so on for all the replicas. If the connection attempt failed for all the replicas, the attempt will be repeated the same way several times.
+
+You can specify up to any number of shards and to any number of replicas for each shard.
```
```suggestion
You can specify any number of shards and number of replicas for each shard.
```

olgarev commented on docs/en/engines/table-engines/integrations/mysql.md:
```diff
@@ -55,6 +55,12 @@ Simple `WHERE` clauses such as `=, !=, >, >=, <, <=` are executed on the MySQL s
 
 The rest of the conditions and the `LIMIT` sampling constraint are executed in ClickHouse only after the query to MySQL finishes.
 
+Supports multiple replicas that must be listed by a character `|`. For example:
```
```suggestion
Supports multiple replicas that must be listed by `|`. For example:
```

olgarev commented on docs/en/engines/table-engines/integrations/postgresql.md:
```diff
@@ -49,6 +49,12 @@ PostgreSQL `Array` types are converted into ClickHouse arrays.
 !!! info "Note"
     Be careful - in PostgreSQL an array data, created like a `type_name[]`, may contain multi-dimensional arrays of different dimensions in different table rows in same column. But in ClickHouse it is only allowed to have multidimensional arrays of the same count of dimensions in all table rows in same column.
+	
+Supports multiple replicas that must be listed by a character `|`. For example:
```
```suggestion
Supports multiple replicas that must be listed by `|`. For example:
```

olgarev commented on docs/en/sql-reference/table-functions/mysql.md:
```diff
@@ -39,6 +39,18 @@ Simple `WHERE` clauses such as `=, !=, >, >=, <, <=` are currently executed on t
 
 The rest of the conditions and the `LIMIT` sampling constraint are executed in ClickHouse only after the query to MySQL finishes.
 
+Supports multiple replicas that must be listed by a character `|`. For example:
```
```suggestion
Supports multiple replicas that must be listed by `|`. For example:
```

olgarev commented on docs/en/sql-reference/table-functions/postgresql.md:
```diff
@@ -43,8 +43,20 @@ PostgreSQL Array types converts into ClickHouse arrays.
 !!! info "Note"
     Be careful, in PostgreSQL an array data type column like Integer[] may contain arrays of different dimensions in different rows, but in ClickHouse it is only allowed to have multidimensional arrays of the same dimension in all rows.
+	
+Supports multiple replicas that must be listed by a character `|`. For example:
```
```suggestion
Supports multiple replicas that must be listed by `|`. For example:
```

olgarev commented on docs/ru/engines/table-engines/integrations/ExternalDistributed.md:
```diff
@@ -0,0 +1,54 @@
+
+Смотрите подробное описание запроса [CREATE TABLE](../../../sql-reference/statements/create/table.md#create-table-query).
+
+Структура таблицы может отличаться от исходной структуры таблицы:
```
```suggestion
Структура таблицы может отличаться от структуры исходной таблицы:
```

olgarev commented on docs/ru/engines/table-engines/integrations/ExternalDistributed.md:
```diff
@@ -0,0 +1,54 @@
+
+Структура таблицы может отличаться от исходной структуры таблицы:
+
+-   Имена столбцов должны быть такими же, как в исходной таблице, но вы можете использовать только некоторые из этих столбцов и в любом порядке.
```
```suggestion
-   Имена столбцов должны быть такими же, как в исходной таблице, но можно использовать только некоторые из этих столбцов и в любом порядке.
```

olgarev commented on docs/ru/engines/table-engines/integrations/ExternalDistributed.md:
```diff
@@ -0,0 +1,54 @@
+
+-   Имена столбцов должны быть такими же, как в исходной таблице, но вы можете использовать только некоторые из этих столбцов и в любом порядке.
+-   Типы столбцов могут отличаться от типов в исходной таблице. ClickHouse пытается [привести](../../../sql-reference/functions/type-conversion-functions.md#type_conversion_function-cast) значения к типам данных ClickHouse.
+-   Настройка `external_table_functions_use_nulls` определяет как обрабатывать Nullable столбцы. Значение по умолчанию: 1. Если значение 0, то табличная функция не будет делать nullable столбцы и будет вместо null выставлять значения по умолчанию для скалярного типа. Это также применимо для null значений внутри массивов.
```
```suggestion
-   Настройка `external_table_functions_use_nulls` определяет как обрабатывать Nullable столбцы. Значение по умолчанию: 1. Если значение 0, то табличная функция не будет делать столбцы Nullable и будет вместо NULL выставлять значения по умолчанию для скалярного типа. Это также применимо для значений NULL внутри массивов.
```

olgarev commented on docs/ru/engines/table-engines/integrations/ExternalDistributed.md:
````diff
@@ -0,0 +1,54 @@
+CREATE TABLE test_shards (id UInt32, name String, age UInt32, money UInt32) ENGINE = ExternalDistributed('MySQL', `mysql{1|2}:3306,mysql{3|4}:3306`, 'clickhouse', 'test_replicas', 'root', 'clickhouse');
+```
+
+При указании реплик для каждого из шардов при чтении будет выбрана одна из доступных реплик. Если соединиться не удалось, то будет выбрана следующая реплика, и так для всех реплик. Если попытка соединения для всех реплик не удалась, то будут снова произведены попытки соединения по кругу и так несколько раз.
````
```suggestion
При указании реплик для каждого из шардов при чтении выбирается одна из доступных реплик. Если соединиться не удалось, то выбирается следующая реплика, и так для всех реплик. Если попытка соединения не удалась для всех реплик, то сервер ClickHouse снова пытается соединиться с одной из реплик, перебирая их по кругу, и так несколько раз.
```

sevirov committed:
Update docs/en/engines/table-engines/integrations/ExternalDistributed.md

Co-authored-by: olgarev <56617294+olgarev@users.noreply.github.com>

sevirov committed:
Update docs/en/engines/table-engines/integrations/ExternalDistributed.md

Co-authored-by: olgarev <56617294+olgarev@users.noreply.github.com>

sevirov committed:
Update docs/en/engines/table-engines/integrations/mysql.md

Co-authored-by: olgarev <56617294+olgarev@users.noreply.github.com>

sevirov committed:
Update docs/en/engines/table-engines/integrations/postgresql.md

Co-authored-by: olgarev <56617294+olgarev@users.noreply.github.com>

sevirov committed:
Update docs/en/sql-reference/table-functions/mysql.md

Co-authored-by: olgarev <56617294+olgarev@users.noreply.github.com>

sevirov committed:
Update docs/en/sql-reference/table-functions/postgresql.md

Co-authored-by: olgarev <56617294+olgarev@users.noreply.github.com>

sevirov committed:
Update docs/ru/engines/table-engines/integrations/ExternalDistributed.md

Co-authored-by: olgarev <56617294+olgarev@users.noreply.github.com>

sevirov committed:
Update docs/ru/engines/table-engines/integrations/ExternalDistributed.md

Co-authored-by: olgarev <56617294+olgarev@users.noreply.github.com>

sevirov committed:
Update docs/ru/engines/table-engines/integrations/ExternalDistributed.md

Co-authored-by: olgarev <56617294+olgarev@users.noreply.github.com>

sevirov committed:
Update docs/ru/engines/table-engines/integrations/ExternalDistributed.md

Co-authored-by: olgarev <56617294+olgarev@users.noreply.github.com>

sevirov committed:
Delete the external_table_functions_use_nulls setting

Удалил настройку external_table_functions_use_nulls из описания движка ExternalDistributed.

sevirov committed:
Replace links

Заменил линки в cast.

sevirov committed:
Update mysql and postgresql engines

Внес небольшие поправки.

sevirov committed:
Merge remote-tracking branch 'remotes/clickhouse/master' into sevirov-DOCSUP-8324-document_the_support_of_replicas_and_shards_in_mysql_postgresql

sevirov committed:
Add links

Добавил ссылку на настройку external_table_functions_use_nulls.

kssenii committed:
Update postgresql.md

kssenii committed:
Update mysql.md

kssenii committed:
Update postgresql.md

kssenii committed:
Update mysql.md

kssenii commented on docs/en/engines/table-engines/integrations/ExternalDistributed.md:
```diff
@@ -0,0 +1,53 @@
+
+# ExternalDistributed {#externaldistributed}
+
+The `ExternalDistributed` engine allows to perform `SELECT` and `INSERT` queries on data that is stored on a remote servers MySQL or PostgreSQL. Accepts [MySQL](../../../engines/table-engines/integrations/mysql.md) or [PostgreSQL](../../../engines/table-engines/integrations/postgresql.md) engines as an argument so sharding is possible.
```
No, the table is read-only.

sevirov committed:
Update ExternalDistributed.md

Убрал поддержку запросов на вставку.
